esp-alloc/src/lib.rs

```diff
@@ -61,35 +67,92 @@ use core::{
 };
 
 use critical_section::Mutex;
+use enumset::{EnumSet, EnumSetType};
 use linked_list_allocator::Heap;
 
+/// The global allocator instance
+#[global_allocator]
+pub static INSTANCE: EspHeap = EspHeap::empty();
+
+const NON_REGION: Option<HeapRegion> = None;
+
+#[derive(EnumSetType)]
+/// Describes the properties of a memory region
+pub enum MemoryCapability {
+    /// Memory must be internal; specifically it should not disappear when
+    /// flash/spiram cache is switched off
+    Internal,
+    /// Memory must be in SPI RAM
+    External,
+}
+
+/// A memory region to be used as heap memory
+pub struct HeapRegion {
+    heap: Heap,
+    capabilities: EnumSet<MemoryCapability>,
+}
+
+impl HeapRegion {
+    /// Create a new [HeapRegion] with the given capabilities
+    pub unsafe fn new(
+        heap_bottom: *mut u8,
+        size: usize,
+        capabilities: EnumSet<MemoryCapability>,
+    ) -> Self {
+        let mut heap = Heap::empty();
+        heap.init(heap_bottom, size);
+
+        Self { heap, capabilities }
+    }
+}
+
+/// For esp-wifi
+#[doc(hidden)]
+#[no_mangle]
+pub extern "C" fn free_internal_heap() -> usize {
+    INSTANCE.free_caps(MemoryCapability::Internal.into())
+}
+
+/// For esp-wifi
+#[doc(hidden)]
+#[no_mangle]
+pub extern "C" fn allocate_from_internal_ram(size: usize) -> *mut u8 {
+    unsafe {
+        INSTANCE.alloc_caps(
+            MemoryCapability::Internal.into(),
+            Layout::from_size_align_unchecked(size, 4),
+        )
+    }
+}
```

Comment:
Does this mean users can override these methods as well if esp-alloc isn't added? (I find the behavior desirable)

Reply:
Yes - this way users can bring their own allocator if they implement these functions

Reply:
Imo, the hard dependency is fine as we can add it behind a feature in esp-wifi.

We can move these functions into esp-wifi (IMO they make more sense there, than hidden here), but behind that esp-alloc feature so that it can be disabled and users can still supply their own allocator and `free_internal_heap` `allocate_from_internal_ram` functions.